### PR TITLE
Add MemoryExtensions.Split

### LINKED ISSUE #76186: [API Proposal]: Alternative span/memory/string splitting API
### Background and motivation

https://github.com/dotnet/runtime/issues/934 has been a long-standing issue about Split support for spans.  It's evolved into an enumerator that wraps IndexOf and Slice into a slightly tidier package.  While that might still be useful, it doesn't make many of the existing use cases for Split very simple, in particular ones where the consumer knows how many split values are expected, wants to extract the Nth value, etc.

Either instead of or in addition to (if we also want an enumerator syntax), we can offer a SplitAsRanges method that operates over `ReadOnlySpan<T>` in a way and stores the resulting ranges into a provided location, that then also works with `Span<T>`, `{ReadOnly}Memory<T>`, and `String`, that doesn't allocate, that automates the retrieval of N values, etc.

### API Proposal

```diff
namespace System;

public static class MemoryExtensions
{
+     public static int SplitAsRanges(this ReadOnlySpan<char> source, Span<Range> destination, char separator, StringSplitOptions options = StringSplitOptions.None);
+     public static int SplitAsRanges(this ReadOnlySpan<char> source, Span<Range> destination, ReadOnlySpan<char> separator, StringSplitOptions options = StringSplitOptions.None);
+     public static int SplitAnyAsRanges(this ReadOnlySpan<char> source, Span<Range> destination, ReadOnlySpan<char> separators, StringSplitOptions options = StringSplitOptions.None);
+     public static int SplitAnyAsRanges(this ReadOnlySpan<char> source, Span<Range> destination, ReadOnlySpan<string> separators, StringSplitOptions options = StringSplitOptions.None);
}
```

- Naming: we could just call these `Spit{Any}`, and have the "ranges" aspect of it be implicit in taking a `Span<Range>` parameter.
- Argument ordering: `destination, separator` vs `separator, destination`?  I went with `destination, separator` so that the configuration-related data (separator and options) are next to each other, but that then does differ from string.Split, where the separator is first.
- The methods all return the number of `System.Range` values written into destination. Use that wants to retrieve N segments regardless of whether there are more can stackalloc a span / allocate an array of N `Range` instance. Use that wants to retrieve N segments and guarantee there are no more than that can stackalloc a span / allocate an array of N+1 `Range` instances, and validate that the returned count was N.
- `System.Range` is `unmanaged` and can be stackalloc'd.
- The stored `Range` instances can be used to slice the original span/memory/string etc. to extract only those values that are needed, in either an allocating or non-allocating manner.

### API Usage

Examples...

1. https://github.com/dotnet/sdk/blob/2f7d6f1928526c29a417c72a9e23497359cbc76f/src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs#L159-L169

Instead of:
```C#
                        string[] dependentParts = dependent.Split(',');

                        if (dependentParts.Length != 3)
                        {
                            Log?.LogMessage($"Skipping dependent: {dependent}");
                            continue;
                        }

                        try
                        {
                            SdkFeatureBand dependentFeatureBand = new SdkFeatureBand(dependentParts[1]);
```
this code could be:
```C#
                        Span<Range> dependentParts = stackalloc Range[4];
                        ...
                        if (dependent.AsSpan().SplitAsRanges(dependentParts, ',') != 3)
                        {
                            Log?.LogMessage($"Skipping dependent: {dependent}");
                            continue;
                        }

                        try
                        {
                            SdkFeatureBand dependentFeatureBand = new SdkFeatureBand(dependent[dependentParts[1]]);
```

2. https://github.com/dotnet/iot/blob/a914669b6b5928d246c0f88486fecc72867bcc76/src/devices/Card/Ndef/Record/GeoRecord.cs#L91-L105

Instead of:
```C#
            var strLatLong = Uri.Substring(4).Split(',');
            if (strLatLong.Length != 2)
            {
                throw new ArgumentException($"Record is not a valid {nameof(GeoRecord)}, can't find a proper latitude and longitude in the payload");
            }

            try
            {
                _latitude = Convert.ToDouble(strLatLong[0], CultureInfo.InvariantCulture);
                _longitude = Convert.ToDouble(strLatLong[1], CultureInfo.InvariantCulture);
            }
            catch (Exception ex) when (ex is FormatException || ex is OverflowException)
            {
                throw new ArgumentException($"Record is not a valid {nameof(GeoRecord)}, can't find a proper latitude and longitude in the payload");
            }
```
this could be:
```C#
            Span<Range> strLatLong = stackalloc Range[3];
            ReadOnlySpan<char> span = Uri.AsSpan(4);
            if (span.Split(strLatLong, ',') != 2)
            {
                throw new ArgumentException($"Record is not a valid {nameof(GeoRecord)}, can't find a proper latitude and longitude in the payload");
            }

            try
            {
                _latitude = double.Parse(span[strLatLong[0]], provider: CultureInfo.InvariantCulture);
                _longitude = double.Parse(span[strLatLong[1]], provider: CultureInfo.InvariantCulture);
            }
            catch (Exception ex) when (ex is FormatException || ex is OverflowException)
            {
                throw new ArgumentException($"Record is not a valid {nameof(GeoRecord)}, can't find a proper latitude and longitude in the payload");
            }
```

3. https://github.com/dotnet/runtime/blob/b0920912512aa7e205363b046a1911e6a97df31f/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.NonAndroid.cs#L86-L118

Instead of:
```C#
                    while ((zoneTabFileLine = sr.ReadLine()) != null)
                    {
                        if (!string.IsNullOrEmpty(zoneTabFileLine) && zoneTabFileLine[0] != '#')
                        {
                            // the format of the line is "country-code \t coordinates \t TimeZone Id \t comments"

                            int firstTabIndex = zoneTabFileLine.IndexOf('\t');
                            if (firstTabIndex >= 0)
                            {
                                int secondTabIndex = zoneTabFileLine.IndexOf('\t', firstTabIndex + 1);
                                if (secondTabIndex >= 0)
                                {
                                    string timeZoneId;
                                    int startIndex = secondTabIndex + 1;
                                    int thirdTabIndex = zoneTabFileLine.IndexOf('\t', startIndex);
                                    if (thirdTabIndex >= 0)
                                    {
                                        int length = thirdTabIndex - startIndex;
                                        timeZoneId = zoneTabFileLine.Substring(startIndex, length);
                                    }
                                    else
                                    {
                                        timeZoneId = zoneTabFileLine.Substring(startIndex);
                                    }

                                    if (!string.IsNullOrEmpty(timeZoneId))
                                    {
                                        timeZoneIds.Add(timeZoneId);
                                    }
                                }
                            }
                        }
                    }
```
this could be:
```C#
                    Span<Range> ranges = stackalloc Range[4];
                    while ((zoneTabFileLine = sr.ReadLine()) != null)
                    {
                        if (zoneTabFileLine.StartsWith('#'))
                        {
                            // the format of the line is "country-code \t coordinates \t TimeZone Id \t comments"
                            int found = zoneTabFileLine.SplitAsRanges(ranges, '\t');
                            if (found >= 3)
                            {
                                timeZoneId = zoneTabFileLine[ranges[3]];
                                if (timeZoneId.Length != 0)
                                {
                                    timeZoneIds.Add(timeZoneId);
                                }
                            }
                        }
                    }
```

### Alternative Designs

https://github.com/dotnet/runtime/issues/934#issuecomment-1165864043

### Risks

_No response_

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/76186

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameParser.cs [hunk 1]
```diff
@@ -190,37 +190,40 @@ private AssemblyNameParts Parse()
 
         private Version ParseVersion(string attributeValue)
         {
-            string[] parts = attributeValue.Split('.');
-            if (parts.Length > 4)
+            ReadOnlySpan<char> attributeValueSpan = attributeValue;
+            Span<Range> parts = stackalloc Range[5];
+            parts = parts.Slice(0, attributeValueSpan.Split(parts, '.'));
+            if (parts.Length is not (3 or 4) ||
+                attributeValue.Contains('\0')) // ushort.TryParse below will ignore trailing nulls, so they're flagged here
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameParser.cs:197]: `attributeValue.Contains('\0')` should not be needed. The nulls are filtered out earlier in `GetNextToken` method.

File: src/libraries/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 2]
```diff
@@ -333,23 +333,23 @@ internal static bool TryFindCGroupPathForSubsystem(CGroupVersion cgroupVersion,
...
+                                path = line[lineParts[2]];
                                 return true;
                             }
                             else if (cgroupVersion == CGroupVersion.CGroup2)
                             {
                                 // cgroup v2: Find the first entry that matches the cgroup v2 hierarchy:
                                 //     0::$PATH
 
-                                if ((lineParts[0] == "0") && (lineParts[1].Length == 0))
+                                if ((lineSpan[lineParts[0]].SequenceEqual("0")) && (lineSpan[lineParts[1]].Length == 0))
```
Comment (Suchiman) [inline src/libraries/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs:350]:
```suggestion
                                if ((lineSpan[lineParts[0]] is "0") && (lineSpan[lineParts[1]].Length == 0))
```
Can be sugared using [Pattern matching on ROS](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/pattern-match-span-of-char-on-string.md)

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/ContentDispositionHeaderValue.cs [hunk 3]
```diff
@@ -470,9 +470,13 @@ private static bool TryDecodeMime(string? input, [NotNullWhen(true)] out string?
...
             }
 
-            string[] parts = processedInput.Split('?');
+            Span<Range> parts = stackalloc Range[6];
+            ReadOnlySpan<char> processedInputSpan = processedInput;
             // "=, encodingName, encodingType, encodedData, ="
-            if (parts.Length != 5 || parts[0] != "\"=" || parts[4] != "=\"" || parts[2].ToLowerInvariant() != "b")
+            if (processedInputSpan.Split(parts, '?') != 5 ||
+                !processedInputSpan[parts[0]].SequenceEqual("\"=") ||
+                !processedInputSpan[parts[4]].SequenceEqual("=\"") ||
```
Comment (Suchiman) [inline src/libraries/System.Net.Http/src/System/Net/Http/Headers/ContentDispositionHeaderValue.cs:478]:
```suggestion
                processedInputSpan[parts[0]] is not "\"=" ||
                processedInputSpan[parts[4]] is not "=\"" ||
```

File: src/libraries/System.Net.Primitives/src/System/Net/CookieContainer.cs [hunk 4]
```diff
@@ -614,13 +614,15 @@ internal bool IsLocalDomain(string host)
             }
 
             // Test for "127.###.###.###" without using regex.
-            string[] ipParts = host.Split('.');
-            if (ipParts != null && ipParts.Length == 4 && ipParts[0] == "127")
+            ReadOnlySpan<char> hostSpan = host;
+            Span<Range> ipParts = stackalloc Range[5];
+            ipParts = ipParts.Slice(0, hostSpan.Split(ipParts, '.'));
+            if (ipParts.Length == 4 && hostSpan[ipParts[0]].SequenceEqual("127"))
```
Comment (Suchiman) [inline src/libraries/System.Net.Primitives/src/System/Net/CookieContainer.cs:620]:
```suggestion
            if (ipParts.Length == 4 && hostSpan[ipParts[0]] is "127")
```

File: src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs [hunk 5]
```diff
@@ -3235,6 +3235,339 @@ private static void SliceLongerSpanToMatchShorterLength<T>(ref ReadOnlySpan<T> s
...
+                    if ((uint)rangeCount >= (uint)destinationMinusOne.Length)
+                    {
+                        break;
+                    }
+
+                    destinationMinusOne[rangeCount] = startInclusive..endExclusive;
+                    rangeCount++;
+                }
+
+                // Reset to be just passed the separator, and loop around to go again.
```
Comment (dakersnar) [inline src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs:3526]:
```suggestion
                // Reset to be just past the separator, and loop around to go again.
```

**Review by dakersnar — APPROVED** [review]

File: src/libraries/Common/src/System/Drawing/ColorConverterCommon.cs [hunk 6]
```diff
@@ -54,26 +54,19 @@ public static Color ConvertFromString(string strValue, CultureInfo culture)
...
+            ReadOnlySpan<char> textSpan = text;
+            Span<Range> tokens = stackalloc Range[5];
+            return textSpan.Split(tokens, sep) switch
             {
-                1 => PossibleKnownColor(Color.FromArgb(values[0])),
-                3 => PossibleKnownColor(Color.FromArgb(values[0], values[1], values[2])),
-                4 => PossibleKnownColor(Color.FromArgb(values[0], values[1], values[2], values[3])),
+                1 => PossibleKnownColor(Color.FromArgb(IntFromString(textSpan[tokens[0]], culture))),
+                3 => PossibleKnownColor(Color.FromArgb(IntFromString(textSpan[tokens[0]], culture), IntFromString(textSpan[tokens[1]], culture), IntFromString(textSpan[tokens[2]], culture))),
+                4 => PossibleKnownColor(Color.FromArgb(IntFromString(textSpan[tokens[0]], culture), IntFromString(textSpan[tokens[1]], culture), IntFromString(textSpan[tokens[2]], culture), IntFromString(textSpan[tokens[3]], culture))),
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Common/src/System/Drawing/ColorConverterCommon.cs:69]: What work is being repeated?

File: src/libraries/System.Private.CoreLib/src/System/Globalization/JapaneseCalendar.Nls.cs [hunk 7]
```diff
@@ -184,29 +184,32 @@ private static int CompareEraRanges(EraInfo a, EraInfo b)
...
             // 3 - Abbreviated English Era Name
-            if (names.Length != 4) return null;
+            Span<Range> names = stackalloc Range[5];
+            ReadOnlySpan<char> dataSpan = data;
+            if (dataSpan.Split(names, '_') == 4)
+            {
+                ReadOnlySpan<char> eraName = dataSpan[names[0]];
+                ReadOnlySpan<char> abbreviatedEraName = dataSpan[names[1]];
+                ReadOnlySpan<char> englishEraName = dataSpan[names[2]];
+                ReadOnlySpan<char> abbreviatedEnglishEraName = dataSpan[names[3]];
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/JapaneseCalendar.Nls.cs:199]: These are extracted as spans, and a span is just a reference and a length.  These are essentially just pointers into the original string.

File: src/libraries/Common/src/System/Drawing/ColorConverterCommon.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dakersnar) [reply src/libraries/Common/src/System/Drawing/ColorConverterCommon.cs:69]: We're converting a given section of textSpan to Int multiple times, right?

[the remark this answers is not in the document]
Comment (dakersnar) [reply src/libraries/Common/src/System/Drawing/ColorConverterCommon.cs:69]: Edit: oh, never mind, just realized these are distinct code paths.

File: src/libraries/Common/src/System/Drawing/ColorConverterCommon.cs [hunk 8]
```diff
@@ -54,26 +54,19 @@ public static Color ConvertFromString(string strValue, CultureInfo culture)
...
-            {
-                values[i] = unchecked(IntFromString(tokens[i], culture));
-            }
-
             // We should now have a number of parsed integer values.
             // We support 1, 3, or 4 arguments:
             //
             // 1 -- full ARGB encoded
             // 3 -- RGB
             // 4 -- ARGB
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Common/src/System/Drawing/ColorConverterCommon.cs]: Fixed

File: src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs (hunk 5 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs]: Fixed

**Review by dakersnar — APPROVED** [review]

**Pull request merged**